storage/block/block.go

```diff
@@ -264,6 +274,52 @@ func (b *dbBlock) ResetRetrievable(
 	b.resetRetrievableWithLock(retriever, metadata)
 }
 
+func (b *dbBlock) stream(ctx context.Context) (xio.SegmentReader, error) {
+	b.ctx.DependsOn(ctx)
+
+	// If the block retrieve ID is set then it must be retrieved
+	var (
+		stream xio.SegmentReader
+		err    error
+	)
+	if b.retriever != nil {
+		start := b.startWithLock()
+		stream, err = b.retriever.Stream(ctx, b.retrieveID, start, b)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		stream = b.opts.SegmentReaderPool().Get()
+		stream.Reset(ts.NewSegment(b.segment.Head, b.segment.Tail, ts.FinalizeNone))
+		ctx.RegisterFinalizer(stream)
+	}
+
+	return stream, nil
+}
+
+// TODO(rartoul): The existing ctx is still holding a reference to the old segment so that will hang around
+// and waste memory until the block is closed. We could improve this by swapping out the underlying ctx with
+// a new one, allowing us to close the old one and release the old segment, freeing memory.
```

Comment:
Do you want to open an issue (so you can close it in the near future 😄 )?

Reply:
#580 tada!